key-revocation.md

```diff
@@ -0,0 +1,76 @@
+
+However, the user must be able to ensure that the key revocation list is accurate and up to date. If an attacker is able to replay an old revocation list, or show different versions to different registries, the user may continue to trust compromised keys. Therefore the distribution of the key revocation list must allow the user to verify authenticity and timeliness.
+
+Also, for security reasons, keys cannot be removed from a key revocation list, so the list will grow larger and larger over time and may eventually have a noticeable bandwidth impact, although this can be mitigated by combining key revocation lists with keys that expire.
+
+Pros:
+* Allows key revocation at any time
+* Anyone can add a key to the revocation list
+
+Cons:
+* Distribution of the key revocation list needs to be secured and verified for timeliness to prevent replays
+* Additional maintenance overhead
+* What to do when the revocation list query fails?
+* Revocation lists must be synchronized between registries.
+
+
+## List of trusted keys (Allow lists)
+
+Instead of distributing untrusted keys, this method distributes a list of currently trusted keys. If a key needs to be revoked, it is removed from the list of trusted keys. This technique has the added benefit of ensuring that users have access to the new trusted key as soon as they learn of a revocation.
+
+Similar to revocation lists, the list of trusted keys must be securely distributed and include mechanisms for the user to verify timeliness.
+
+Pros:
+* Allows key revocation at any time
+* Can be combined with key distribution
+* Does not require an additional query
+
+Cons:
+* Distribution of trusted key list needs to be secured and verified for timeliness to prevent replays
+* Additional maintenance overhead
+
+
+## Combining explicit and implicit revocation
+
+A final option is to use a combination of the first and third techniques to achieve both implicit and explicit key revocation. By using a hierarchical combination of keys, a trusted root key can delegate signing to various keys that expire. These keys may be revoked before that expiration time by replacing the key listed in the delegation. Additionally, artifacts may be signed by more than one key, allowing automated tooling to provide short lived signatures that verify the signer and artifact have not been revoked. Clients then verify the necessary collection of signatures is found on the artifact.
+
+This method allows signers to have relatively long lived keys, to simplify their workflow and avoid needing to resign the artifacts themselves, while enabling timely revoking of the signing key or a single artifact signature.
+
+For efficiency, a meta-artifact can be created and maintained, containing references to a collection currently signed artifacts. And the short lived signature can be created for this single artifact, rather than every artifact individually.
```

Comment:
How would this interact with individual artifacts moving within and across registries?

Reply:
it depends on the implementation, but the meta-artifact would have to be updated as the collection changes.